### ISSUE TITLE
[Question]  Background removal question.

### ISSUE THREAD

**Issue by ovalerio** [CAREamics/careamics#799]
## Should I remove the  background of the raw data before training n2v model?

What are your recommendation/guidelines regarding background removal in the context of training a n2v model using light microscopy images.

The images I am training the denoise model for contain sparse bright structures (moving worms) and some halo artifacts.

**Comment by jdeschamps** [comment]
Hi @ovalerio,

Do you have the feeling that Noise2Void is sub-optimal with your images?

We didn't often get data with bg_area >> signal_area, but in one instance in the past Noise2Void did not even train properly due to the overwhelming presence of background. The solution was to pre-process the data into patches and filter out most background patches, then use the filtered patches as training data. The resulting model could be used on the whole images.

We are currently putting a lot of effort in trying to finalize `v0.1.0`, which brings about a much more powerful dataset class. In particular background patch filtering is one of the new feature. It is already usable via the Lightning API. But it is not yet documented. It is quite inefficient, and current work (as we speak) is being carried out in writing a better implementation (probably merged next week).

If you are interested in testing it, I would advise to wait for the new implementation. In the mean time, you can pre-process your images. We have several simple routines to detect background patches. However, they are not parameter-free and will require some parameter tuning.

The way our background filtering works is with two features (all internal to the new dataset):
- Provide a mask image that will be used to reject patches
- or filter image areas via a first pass through the data using particular statistics-based routines

The thing is, for your current problem, is that these statistics-based routines can also be used outside of our dataset on your images to exclude background area.

The statistics-based routines to detect background patches I mentioned above are implemented in [here](https://github.com/CAREamics/careamics/tree/main/src/careamics/dataset_ng/patch_filter) (`max_filter.py`, `mean_std_filter.py`, and `shannon_filter.py`). Each of these implementation has a `filter_map` method that you can call on your image to obtain a map of potential areas with signal. Each implementation has a complete example in the docstring. For instance for the `max_filter.py` module:

https://github.com/CAREamics/careamics/blob/9730fce6c0bd0efc44c2718194a36d83f3e89561/src/careamics/dataset_ng/patch_filter/max_filter.py#L111-L124

You can try this on your data. The resulting `map` needs to be thresholded similarly to how the filter works. For the example above, this is done on line 122:

https://github.com/CAREamics/careamics/blob/9730fce6c0bd0efc44c2718194a36d83f3e89561/src/careamics/dataset_ng/patch_filter/max_filter.py#L122

The result of the thresholding is a mask you will be able to use in the future with our dataset (or alternatively you use the best `threshold` in the configuration to do the automatic filtering), or you can already use the mask to filter your data yourself and build a training set of patches.

Note that the halo (if I understand what you mean by it) is probably signal and there is not much to do about it in the context of denoising.

Let me know if things are not clear, or you have other questions. :)
